### PR TITLE
[GCP UPI] remove bootstrap ig from backend service of random id

### PR DESCRIPTION
**Pull request by jianli-wei**
debug the GCP UPI failure [periodic-ci-openshift-openshift-tests-private-release-4.17-multi-nightly-gcp-upi-private-xpn-ingress-glb-arm-f14](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/logs/periodic-ci-openshift-openshift-tests-private-release-4.17-multi-nightly-gcp-upi-private-xpn-ingress-glb-arm-f14/1823736568330850304)
```
2024-08-14 15:40:38+00:00 - Bootstrap complete, destroying bootstrap resources
[DEBUG] Running Command: 'gcloud compute backend-services remove-backend ci-op-rz922n76-3da48-6sfgf-api-internal --region=us-central1 --instance-group=ci-op-rz922n76-3da48-6sfgf-bootstrap-ig --instance-group-zone=us-central1-a'
Updated [https://www.googleapis.com/compute/v1/projects/XXXXXXXXXXXX/regions/us-central1/backendServices/ci-op-rz922n76-3da48-6sfgf-api-internal].
Removing gs://ci-op-rz922n76-3da48-6sfgf-bootstrap-ignition/bootstrap.ign...
/ [1 objects]                                                                   
Operation completed over 1 objects.                                              
Removing gs://ci-op-rz922n76-3da48-6sfgf-bootstrap-ignition/...
Waiting for delete [operation-1723650085823-61fa68f1523e4-20b9c981-cde2075d]...
......................failed.
ERROR: (gcloud.deployment-manager.deployments.delete) Delete operation operation-1723650085823-61fa68f1523e4-20b9c981-cde2075d failed.
Error in Operation [operation-1723650085823-61fa68f1523e4-20b9c981-cde2075d]: errors:
- code: RESOURCE_ERROR
  location: /deployments/ci-op-rz922n76-3da48-6sfgf-bootstrap/resources/ci-op-rz922n76-3da48-6sfgf-bootstrap-ig
  message: "{\"ResourceType\":\"compute.v1.instanceGroup\",\"ResourceErrorCode\":\"\
    400\",\"ResourceErrorMessage\":{\"code\":400,\"errors\":[{\"domain\":\"global\"\
    ,\"message\":\"The instance_group resource 'projects/XXXXXXXXXXXX/zones/us-central1-a/instanceGroups/ci-op-rz922n76-3da48-6sfgf-bootstrap-ig'\
    \ is already being used by 'projects/XXXXXXXXXXXX/regions/us-central1/backendServices/aebe62e5c12494948ba956f960842d22'\"\
    ,\"reason\":\"resourceInUseByAnotherResource\"}],\"message\":\"The instance_group\
    \ resource 'projects/XXXXXXXXXXXX/zones/us-central1-a/instanceGroups/ci-op-rz922n76-3da48-6sfgf-bootstrap-ig'\
    \ is already being used by 'projects/XXXXXXXXXXXX/regions/us-central1/backendServices/aebe62e5c12494948ba956f960842d22'\"\
    ,\"statusMessage\":\"Bad Request\",\"requestPath\":\"https://compute.googleapis.com/compute/v1/projects/XXXXXXXXXXXX/zones/us-central1-a/instanceGroups/ci-op-rz922n76-3da48-6sfgf-bootstrap-ig\"\
    ,\"httpMethod\":\"DELETE\"}}"

### CODE REVIEW

**Comment by jianli-wei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-multi-nightly-gcp-upi-private-xpn-ingress-glb-arm-f14

**Comment by jianli-wei** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-gcp-upi-private-xpn-ingress-glb-arm-f14

**Comment by jianli-wei** [comment]
FYI
The 4.17 behavior (see [test](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/55605/rehearse-55605-periodic-ci-openshift-openshift-tests-private-release-4.17-multi-nightly-gcp-upi-private-xpn-ingress-glb-arm-f14/1823989308810334208)):
```
2024-08-15 08:19:22+00:00 - Bootstrap complete, destroying bootstrap resources
[DEBUG] Running Command: 'gcloud compute backend-services remove-backend ci-op-cp39i983-31e04-nb75c-api-internal --region=us-central1 --instance-group=ci-op-cp39i983-31e04-nb75c-bootstrap-ig --instance-group-zone=us-central1-b'
Updated [https://www.googleapis.com/compute/v1/projects/XXXXXXXXXXXX/regions/us-central1/backendServices/ci-op-cp39i983-31e04-nb75c-api-internal].
[DEBUG] the backend service of random id: 'a715c35b94c954986836f7a0c3bb7d51'
[DEBUG] Running Command: 'gcloud compute backend-services describe a715c35b94c954986836f7a0c3bb7d51 --region us-central1 --format json | jq -r .backends | grep "ci-op-cp39i983-31e04-nb75c-bootstrap-ig"'
    "group": "https://www.googleapis.com/compute/v1/projects/XXXXXXXXXXXX/zones/us-central1-b/instanceGroups/ci-op-cp39i983-31e04-nb75c-bootstrap-ig"
[DEBUG] Running Command: 'gcloud compute backend-services remove-backend a715c35b94c954986836f7a0c3bb7d51 --region=us-central1 --instance-group=ci-op-cp39i983-31e04-nb75c-bootstrap-ig --instance-group-zone=us-central1-b'
Updated [https://www.googleapis.com/compute/v1/projects/XXXXXXXXXXXX/regions/us-central1/backendServices/a715c35b94c954986836f7a0c3bb7d51].
Removing gs://ci-op-cp39i983-31e04-nb75c-bootstrap-ignition/bootstrap.ign...
/ [1 objects]                                                                   
Operation completed over 1 objects.                                              
Removing gs://ci-op-cp39i983-31e04-nb75c-bootstrap-ignition/...
Waiting for delete [operation-1723710056367-61fb4859b068b-c4383382-c263f4f2]...
.........................done.
Delete operation operation-1723710056367-61fb4859b068b-c4383382-c263f4f2 completed successfully.
```
The 4.16 behavior (see [test](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/55605/rehearse-55605-periodic-ci-openshift-openshift-tests-private-release-4.16-multi-nightly-gcp-upi-private-xpn-ingress-glb-arm-f14/1823999430647353344)):
```
2024-08-15 08:57:35+00:00 - Bootstrap complete, destroying bootstrap resources
[DEBUG] Running Command: 'gcloud compute backend-services remove-backend ci-op-z9l0x00f-b347d-9q6hz-api-internal-backend-service --region=us-central1 --instance-group=ci-op-z9l0x00f-b347d-9q6hz-bootstrap-ig --instance-group-zone=us-central1-f'
Updated [https://www.googleapis.com/compute/v1/projects/XXXXXXXXXXXX/regions/us-central1/backendServices/ci-op-z9l0x00f-b347d-9q6hz-api-internal-backend-service].
[DEBUG] the backend service of random id: 'a5b7d571bd14e4d64aeb420cb7dc56e7'
[DEBUG] Running Command: 'gcloud compute backend-services describe a5b7d571bd14e4d64aeb420cb7dc56e7 --region us-central1 --format json | jq -r .backends | grep "ci-op-z9l0x00f-b347d-9q6hz-bootstrap-ig"'
[DEBUG] ci-op-z9l0x00f-b347d-9q6hz-bootstrap-ig is not in a5b7d571bd14e4d64aeb420cb7dc56e7 backends.
Removing gs://ci-op-z9l0x00f-b347d-9q6hz-bootstrap-ignition/bootstrap.ign...
/ [1 objects]                                                                   
Operation completed over 1 objects.                                              
Removing gs://ci-op-z9l0x00f-b347d-9q6hz-bootstrap-ignition/...
Waiting for delete [operation-1723712309923-61fb50bed8e63-e97ff0c8-d52fda00]...
.....................done.
Delete operation operation-1723712309923-61fb50bed8e63-e97ff0c8-d52fda00 completed successfully.

**Comment by jianli-wei** [comment]
@jianlinliu @gpei please help review, thanks!

**Comment by jianlinliu** [comment]
Is this behavior is introduced by 4.17 CAPI ? If yes, shall we file a doc bug for the UPI install doc ?

**Comment by jianli-wei** [comment]
> Is this behavior is introduced by 4.17 CAPI ? If yes, shall we file a doc bug for the UPI install doc ?

Yes, I think it's introduced by 4.17 CAPI. Let me confirm and file a doc bug. Thanks!

**Review by barbacbd — COMMENTED** [review]
/lgtm

**Comment by jianlinliu** [comment]
cc @liangxia to review.

**Comment by jianlinliu** [comment]
/pj-rehearse pull-ci-openshift-installer-release-4.14-e2e-gcp-ovn-upi

**Comment by jianlinliu** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.14-e2e-gcp-ovn-upi

File: ci-operator/step-registry/upi/install/gcp/upi-install-gcp-commands.sh [hunk 1]
```diff
@@ -601,6 +601,20 @@ else # for workflow before internal load balancers
   gcloud compute target-pools remove-instances "${INFRA_ID}-ign-target-pool" "--instances-zone=${ZONE_0}" "--instances=${INFRA_ID}-bootstrap"
   gcloud compute target-pools remove-instances "${INFRA_ID}-api-target-pool" "--instances-zone=${ZONE_0}" "--instances=${INFRA_ID}-bootstrap"
 fi
+# remove from the backend service of random id
+backend_service_random_id=$(gcloud compute backend-services list --format=json --filter="backends[].group~${INFRA_ID}" | jq -r .[].name | grep -v "${INFRA_ID}")
```
Comment (jianlinliu) [inline ci-operator/step-registry/upi/install/gcp/upi-install-gcp-commands.sh:605]: When `backend_service_random_id` is empty, the command will return non-zero, the line will cause the whole process exit as failure. Refer to https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/55605/rehearse-55605-periodic-ci-openshift-release-master-nightly-4.14-e2e-gcp-ovn-upi/1824293917588197376

**Comment by jianli-wei** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.14-e2e-gcp-ovn-upi

**Comment by jianli-wei** [comment]
The UPI installation succeeds. 

Test name | Commit | Details | Comments
-- | -- | -- | --
ci/rehearse/periodic-ci-openshift-release-master-nightly-4.14-e2e-gcp-ovn-upi | https://github.com/openshift/release/commit/7f965f8c14d0d136d879f4e032c48b0943cc2c87 | [link](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/55605/rehearse-55605-periodic-ci-openshift-release-master-nightly-4.14-e2e-gcp-ovn-upi/1824355697811263488) | INFO[2024-08-16T08:34:42Z] Step phase pre succeeded after 31m30s

**Comment by jianli-wei** [comment]
@jianlinliu please help review again, thanks!

**Pull request merged**